docs: eliminate unnecessary information from first contribution page

holmanb opened the pull request:
<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [ ] I have signed the CLA: https://ubuntu.com/legal/contributors
- [ ] I have included a comprehensive commit message using the guide below
- [ ] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [ ] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [ ] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [ ] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e doc``.
-->


## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->
```
docs: eliminate duplicate information from first PR page

First time contributors primarily need to understand what is expected in a
contribution. This creates a self-contained page that explains the bulk of
those expectations in a streamlined fashion that minimizes the amount of
additional unnecessary information. It also eliminates some redundancy,
and removes the unnecessarily verbose code review page.
```

## Context

Expectations regarding contributions were distributed throughout various pages - in some cases intermixed with unrelated information.

The ideas communicated in the code review page could be communicated to contributors in far simpler terms.

This PR addresses both.

### Changed files



#### new files

  dev_docs.rst -> container for non-user-facing details

#### renamed files

  packaging.rst -> build_system.rst
  package_testing.rst -> build_distro_packaging_from_source.rst
  first_PR.html -> development/index.html 

#### split files

  testing.rst -> howto/test_unreleased_packages.rst

#### deleted files

  contribute_testing.rst

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->


## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

TheRealFalcon reviewed (commented):
Nice additions, but I'm -1 to most of the changes.

TheRealFalcon commented:
I do wonder if it's worth putting a consolidated version of some of this info in the top-level README of the github project. Most projects have some sort of "getting started" section in their README that includes how to run the code and run tests. It could just be some kind of statement about how running the code requires some work with a link to the tutorial, and any other operations involve running `tox`.

holmanb commented on doc/rtd/development/first_PR.rst:
```diff
@@ -1,55 +1,62 @@
-cloud-init. Please refer to the `GitHub quickstart`_ documentation
-for more information.
+Make sure that you understand the
+:ref:`code review process<code_review_process>`.
```
Agreed. I don't think that it adds much value, except maybe for those that might have never contributed to open source before. I extracted some of the details that were unique to that page and removed that file.

holmanb commented:
Thanks for the early feedback @TheRealFalcon. I addressed your comments, and also clarified the intent in the proposed commit message in the description.

holmanb commented:
> As a contributor, I would expect that after cloing the project I can just run make and that the README in the project would also explain some basics.

The README points to the contribution docs, which should have the information required to get started as a contributor. It isn't a valuable use of developer time to maintain duplicates of the same information. Make isn't a universal tool that is always used by developers. As someone who regularly contributes to various projects, I can tell you that every project has different contribution expectations, and the best way to jump into a project is to first read the contribution docs. This project is no different - intuition might work sometimes, but sometimes it won't.

holmanb commented on doc/rtd/development/first_PR.rst:
```diff
@@ -1,55 +1,113 @@
-:ref:`our QEMU tutorial<tutorial_qemu>`, which walks through this process
-step-by-step.
+Modify the source: fix a bug or add a feature. Make sure to update comments
+and docs, limit unnecessary code changes, and remove any code that is no longer
```
s/limit/avoid/

blackboxsw commented on doc/rtd/development/index.rst:
```diff
@@ -1,83 +1,35 @@
-* Format code (using ``black`` and ``isort``) with ``tox -e do_format``.
-* Ensure unit tests and/or linting checks pass using ``tox``.
-* Submit a PR against the ``main`` branch of the cloud-init repository.
+Make sure to read the `first PR guide<first-PR>` before starting
```
1. We've redacted this index to something that only drives contributors to code contribution via first-PR, and omits bug filing, docs or testing contributions.

Can we retain cross-links to the other contribution scopes otherwise this index page feels spartan and leaves the other nested menu items form the left-side navigation without much context.


* Contribute code
* Contribute to documentation
* Contribute testing
* Contribute bugs or feature requests



2. Also, let's fix the typo resulting in bogus sphinx anchor
```suggestion
Make sure to read the `first PR guide<first_PR>` before starting
```

holmanb commented on README.md:
```diff
@@ -22,37 +22,22 @@ access key and many other aspects of a system. Later on cloud-init will
 The majority of [clouds](https://docs.cloud-init.io/en/latest/reference/datasources.html#datasources_supported)
 and [Linux / Unix OSes](https://docs.cloud-init.io/en/latest/reference/distros.html)
 are supported by and ship with cloud-init. If your distribution or cloud is not
 supported, please get in contact with that distribution and send them our way!
```
It doesn't really make sense to have these links to refrence data be stored between "Getting help" and "Start developing", which are both actions that one might take. It makes more sense combined up into the section above which describes cloud-init as a project.

holmanb commented on doc/rtd/development/build_system.rst:
```diff
@@ -1,7 +1,7 @@
 
-Downstream packaging
-********************
+Build system
```
This page is actually about the build system used by cloud-init, not about downstream packaging. Name it accordingly.

holmanb commented on doc/rtd/development/code_review.rst:
```diff
@@ -1,236 +0,0 @@
-.. _code_review_process:
```
This page had a couple of useful tidbits and a lot of words to essentially describe the state machine of open source contributions. Extract the interesting bits and retire the page.

holmanb commented on doc/rtd/development/contribute_code.rst:
```diff
@@ -62,21 +44,6 @@ version changed:
-annotations in the style specified by `PEP-484`_ and `PEP-526`_. Their use in
-the codebase is encouraged.
-
-
```
this is mentioned in development/index.rst

holmanb commented on doc/rtd/development/contribute_docs.rst:
```diff
@@ -49,14 +50,6 @@ layout of the ``doc`` directory in the :doc:`docs_layout` article.
-If you need some help with your contribution, you can contact us on our
-`Matrix room <Matrix_>`_. If you have already submitted a
-work-in-progress PR, you can also ask for guidance from our technical author
-by `tagging s-makin`_ as a reviewer.
```
This information is dated and duplicate of other information. Drop it.

holmanb commented on doc/rtd/development/first_PR.rst:
```diff
@@ -1,55 +0,0 @@
-Submit your first pull request
```
The new index.rst is based on this page.

holmanb commented on doc/rtd/howto/bugs.rst:
```diff
@@ -46,17 +54,25 @@ bug:
 
-Distro-specific issues
-======================
+Report documentation bugs
```
This came from finding_issues.rst. This seems to have been misplaced.

holmanb commented on doc/rtd/index.rst:
```diff
@@ -95,14 +97,8 @@ projects, contributions, suggestions, fixes and constructive feedback.
-   Contribute to docs <development/contribute_docs.rst>
-   Contribute to testing <development/contribute_testing.rst>
-   Community <development/summit.rst>
-   Downstream packaging <development/packaging.rst>
```
This is developer documentation. Hide these pages from user distraction under a deeper taxonomy that encourages developers to first read `index.html`.

holmanb commented on doc/rtd/index.rst:
```diff
@@ -62,25 +61,28 @@ to the tool.
   ``cloud-init``
-- :ref:`Try the FAQ<faq>` for answers to some common questions
-- You can also check `GitHub Discussions`_
-- Find a bug? `Report bugs on GitHub Issues`_
```
Direct bug reporters to the "how to file a bug guide" instead.

holmanb commented on doc/rtd/reference/faq.rst:
```diff
@@ -64,12 +53,6 @@ Blog Posts:
 - `Setup Neovim cloud-init Completion`_
   [neovim, yaml, Language Server Protocol, jsonschema, cloud-init]
 
-Events:
```
redundant

blackboxsw commented on doc/rtd/development/index.rst:
```diff
@@ -1,38 +1,135 @@
+   Read dev docs <dev_docs.rst>
 
-.. LINKS:
+.. _how to use Git and GitHub: _https://docs.github.com/en/get-started/start-your-journey
```
Broken link. I think it may be nice for our CI workflow to trigger `.github/workflows/linkcheck.yml` to also trigger on branches we know are documentation-related branches.  Currently, they are only scheduled as a CRON job. It doesn't need to be this PR so I've created a [low-hanging fruit enhancement issue for CI](https://github.com/canonical/cloud-init/issues/6594).
```suggestion
.. _how to use Git and GitHub: https://docs.github.com/en/get-started/start-your-journey
```

blackboxsw reviewed (approved):
I like your new categorization of dev_docs to provide more structure around developer needs. It clarifies the messaging a bit and the intent of this area of our documentation.

Minor suggestions throughout  the only icky things are:
1. the duplicates issues link in doc/rtd/howto/bugs.rst. Feel free to resolve that either local to howto/bugs.rst  or more broadly via doc/rtd/links.txt
2. broken link to github with leading underscore


I think we are missing another doc file rename in this changeset. We need to be a bit careful about those renames as they generate 404s on RTD.com/docs.cloud-init.io etc that require admin account redirects for our project. 

To ensure we better track file removals and renames, let's make sure we represent all doc filename changes  in the PR description as you have

What's missing:
- first_PR.html -> development/index.html maybe?

After this lands I'll update rtd.com/cloud-init admin settings to ensure the proper redirects are added just in case external sites are linking some of these docs.
